# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [4]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [5]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [8]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [9]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [10]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [11]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [12]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [13]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [14]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [15]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [16]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [17]:
data.duplicated().sum()

54

In [18]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [19]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [37]:
#ch_d = data[['debt', 'children']] #создаем отдельную таблицу с количеством детей и показателем задолженности по возврату кредитов
#ch_total=ch_d.groupby('children').count()
#print(ch_total.sort_values(by='debt',ascending=False)) #подсчитываем количество клиентов с определенным количеством детей для наглядности
#ch_d = ch_d.loc[ch_d['debt'] == 1] #сужаем данные до клиентов с просрочками
#ch_negative=ch_d.groupby('children').count() #расчет количества клиентов с просрочками с учетом количества детей
#print(ch_negative.sort_values(by='debt',ascending=False))
#table0=100*ch_negative/ch_total #вычисляем долю в % клиентов с просрочками
#print(table0.sort_values(by='debt',ascending=False))
debt_grouped = data.groupby('children').agg({'debt':['count','sum']})#посчитаем общее количество людей для каждой категории по количеству детей и количество людей, имеющих задолженность по этой категории
debt_grouped['percent']=round((debt_grouped['debt']['sum']/debt_grouped['debt']['count']*100),2) #узнаем, какой процент от общего количества людей с определенным количеством детей имеет задолженность
print(debt_grouped.sort_values(by='percent',ascending=False))



           debt       percent
          count   sum        
children                     
4            41     4    9.76
2          2052   194    9.45
1          4809   444    9.23
3           330    27    8.18
0         14107  1063    7.54
5             9     0    0.00


**Вывод:** лучший показатель своевременности возврата кредита среди широких категорий клиентов у потребителей без детей, причем с отрывом в 0,6% от категории клиентов с 3 детьми. Клиенты с 5 детьми вовсе не имеют просрочек по кредитам, однако и представлены в самой мизерной по выборке категории из 9 потребителей. Худший показатель у клиентов с 4 детьми - 9,76%, однако выборка здесь так же не велика - только 41 человек.
В исходной выборке объемы клиентов с разным количеством детей отличаются в разы. Например, клиентов с 1 ребнком 4809, а с двумя детьми - 2052 (что в 2,3 раза меньше), при этом просрочки равны 444 и 194 соответственно (что так же в 2,3 раза меньше). По остальным парам ситуация схожая. 
Объем просрочек кредита в количественном выражении снижается с ростом детей, потому что самих клиентов с большим количеством детей меньше в разы, чем бездетных или немногодетных. Зависимость между количеством детей и возвратом кредита в срок сомнительна.

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [38]:
#fs_d = data[['debt', 'family_status']] #создаем отдельную таблицу с семейным положением и показателем задолженности по возврату кредитов
#Альтернативное решение:
#print(fs_d.groupby('family_status').count()) #подсчитываем количество клиентов с определенным семейным положением для наглядности
#fs_d = fs_d.loc[fs_d['debt'] == 1] #сужаем данные до клиентов с просрочками
#print(fs_d.groupby('family_status').count()) #группируем по семейному положению объем клиентов с просрочками

#fs_total=fs_d.groupby('family_status').count() #подсчитываем количество клиентов с определенным семейным положением для наглядности
#print(fs_total.sort_values(by='debt',ascending=False))
#fs_d = fs_d.loc[fs_d['debt'] == 1] #сужаем данные до клиентов с просрочками
#fs_negative=fs_d.groupby('family_status').count() #расчет количества клиентов с просрочками по каждому виду семейного положения
#print(fs_negative.sort_values(by='debt',ascending=False))
#table1=100*fs_negative/fs_total #вычисляем долю в % клиентов с просрочками
#print(table1.sort_values(by='debt',ascending=False)) #сортируем по семейному положению долю в % клиентов с просрочками по убыванию

debt_grouped_1 = data.groupby('family_status').agg({'debt':['count','sum']})#посчитаем общее количество людей для каждой категории по семейному положению и количество людей, имеющих задолженность по этой категории
debt_grouped_1['percent']=round((debt_grouped_1['debt']['sum']/debt_grouped_1['debt']['count']*100),2) #узнаем, какой процент от общего количества людей с определенным семейным положением имеет задолженность
print(debt_grouped_1.sort_values(by='percent',ascending=False))


                        debt      percent
                       count  sum        
family_status                            
Не женат / не замужем   2796  273    9.76
гражданский брак        4146  385    9.29
женат / замужем        12266  927    7.56
в разводе               1189   84    7.06
вдовец / вдова           951   63    6.62


**Вывод:** зависимость между семейным положением и возвратом кредита в срок несовсем очевидная. Наивысший риск просрочки у группы "Не женат / не замужем", хотя это лишь 3-я по объему выборки группа, а также высокий показатель у лиц, состоящих в гражданском браке, женатые клиенты, хоть и представляют самую крупную группу в выборке, идут следом с разрывом в 1,7%. Наиболее привлекательными для банка клиентами являются те, кто имеют статус "вдовец/вдова" и "в разводе".

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [41]:
#ig_d = data[['debt', 'total_income_category']] #создаем отдельную таблицу с уровнем дохода и показателем задолженности по возврату кредитов

#ig_total=ig_d.groupby('total_income_category').count() #подсчитываем количество клиентов с определенным уровнем дохода для наглядности
#print(ig_total.sort_values(by='debt',ascending=False))
#ig_d = ig_d.loc[ig_d['debt'] == 1] #сужаем данные до клиентов с просрочками
#ig_negative=ig_d.groupby('total_income_category').count() #расчет количества клиентов с просрочками по каждому уровню дохода
#print(ig_negative.sort_values(by='debt',ascending=False))
#table2=100*ig_negative/ig_total #вычисляем долю в % клиентов с просрочками
#print(table2.sort_values(by='debt',ascending=False)) #сортируем по уровню дохода долю в % клиентов с просрочками по убыванию

#0–30000 — 'E'; 30001–50000 — 'D'; 50001–200000 — 'C'; 200001–1000000 — 'B'; 1000001 и выше — 'A'.
debt_grouped_2 = data.groupby('total_income_category').agg({'debt':['count','sum']})#посчитаем общее количество людей для каждой категории по уровню дохода и количество людей, имеющих задолженность по этой категории
debt_grouped_2['percent']=round((debt_grouped_2['debt']['sum']/debt_grouped_2['debt']['count']*100),2) #узнаем, какой процент от общего количества людей с определенным доходом имеет задолженность
print(debt_grouped_2.sort_values(by='percent',ascending=False))

                        debt       percent
                       count   sum        
total_income_category                     
E                         22     2    9.09
C                      15938  1353    8.49
A                         25     2    8.00
B                       5014   354    7.06
D                        349    21    6.02


**Вывод:** В данной выборке клиентов с самым низким и самым высоким уровнем дохода примерно равное и малое количество, в отличие от других групп доходов, поэтому серьезно оценивать их корреляцию нерелевантно. Как бы то ни было, наименьшей привлекательностью для банка обладает клиентская группа Е, причем ближайшая по уровню дохода группа D имеет лучший показатель возврата кредита в срок с отрывом от почти самой "богатой" группы В - в 1%. Клиенты с самым высоким уровнем дохода имеют средний показатель возврата кредита, в то время как группа клиентов с уровнем дохода 50001–200000 максимально приближена к худшему показателю и отстает от Е на 0,6%.

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

In [43]:
#pu_d = data[['debt', 'purpose_category']] #создаем отдельную таблицу с целью кредита и показателем задолженности по возврату кредитов

#pu_total=pu_d.groupby('purpose_category').count() #подсчитываем количество клиентов с определенной целью кредита
#print(pu_total.sort_values(by='debt',ascending=False))
#pu_d = pu_d.loc[pu_d['debt'] == 1] #сужаем данные до клиентов с просрочками
#pu_negative=pu_d.groupby('purpose_category').count() #расчет количества клиентов с просрочками по каждой цели кредита
#print(pu_negative.sort_values(by='debt',ascending=False))
#table3=100*pu_negative/pu_total #вычисляем долю в % клиентов с просрочками
#print(table3.sort_values(by='debt',ascending=False)) #сортируем по цели кредита долю в % клиентов с просрочками по убыванию

debt_grouped_3 = data.groupby('purpose_category').agg({'debt':['count','sum']})#посчитаем общее количество людей для каждой категории по целям кредита и количество людей, имеющих задолженность по этой категории
debt_grouped_3['percent']=round((debt_grouped_3['debt']['sum']/debt_grouped_3['debt']['count']*100),2) #узнаем, какой процент от общего количества людей с определенной целью кредита имеет задолженность
print(debt_grouped_3.sort_values(by='percent',ascending=False))


                           debt      percent
                          count  sum        
purpose_category                            
операции с автомобилем     4281  400    9.34
получение образования      3989  369    9.25
проведение свадьбы         2324  183    7.87
операции с недвижимостью  10754  780    7.25


**Вывод:** Больше всего кредитов клиенты берут на операции с недвижимостью, при этом по данной цели лучший показатель своевременности возврата. По цели "операции с автомобилем" ситуация в корне обратная, кредиты, выданные на данную цель, наиболее рискованные для банка. Показатель просрочки по кредитам в образовательных целях максимально приближен к худшему. На "проведение свадьбы" выделяется меньше всего кредитов и показатель своевременности погашения здесь приближен к лучшему.

In [28]:
# Код ревьюера:

debt_grouped = data.groupby('children').agg({'debt':['count','sum']})#посчитаем общее количество людей для каждой категории по количеству детей и количество людей, имеющих задолженность по этой категории
debt_grouped['percent']=round((debt_grouped['debt']['sum']/debt_grouped['debt']['count']*100),2) #узнаем, какой процент от общего количества людей с определенным количеством детей имеет задолженность
print(debt_grouped)

           debt       percent
          count   sum        
children                     
0         14107  1063    7.54
1          4809   444    9.23
2          2052   194    9.45
3           330    27    8.18
4            41     4    9.76
5             9     0    0.00


**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*Ответ:* В исходных данных были обнаружены пропуски в ежемесячных доходах и трудовом стаже. В первую очередь, это связано с тем, что многие клиенты не хотят раскрывать данную информацию, считая ее максимально личной и конфиденциальной. Во-вторых, клиенты на момент оформления кредита могли не знать точный трудовой стаж и пропустить данную графу. В третьих, могла произойти потеря данных ввиду человеческого фактора (по вине сотрудника банка), либо ввиду технической ошибки при переносе или сохранении данных.

<div class="alert-success"> 
<b>Комментарий ревьюера 👍 </b>
    
Да, это действительно наиболее возможные (и частые) причины появления пропусков

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:*  <div class="alert-info"> Медиана — это число, которое окажется строго по центру списка чисел в наборе данных, если их предварительно упорядочить по возрастанию. При заполнении пропусков в количественных переменных - это лучшее решение, так как оно не создаст сильного разрыва в данных, придав скученность показателям в районе срединного уровня. Если значение выброса увеличивается, то медиана не изменяется, в тов ремя как среднее значение растет, реагируя чувствительнее. Преддположим, если произойдет изъятие наименьшего значения, то медиана и среднее увеличится, но среднее вырастет сильнее.

## Общий вывод.

Таким образом, удалось выявить, что наиболее надежный заемщик соответствует следующему паттерну: вдова или вдовец (человек в разводе) с ежемесячным доходом 30001–50000 (или 200001–1000000), не имеющий детей и берущий кредит на операции с недвижимостью (или проведение свадьбы).

Наименее надежный заемщик соответствует следующему паттерну: человек не женат/не замужем, с 2 или 4 детьми, с ежемесячным доходом 0–30000 (или 50001–200000), берущий кредит на операции с автомобилем  (или получение образования).